apps/backend/app/services/expression_evaluator.py

```
@@ -0,0 +1,679 @@
+        raise ExpressionError(
+            f"Unknown operation type: {op_type}",
+            node_id=node.node_id
+        )
+
+    def _eval_logical_and(self, node: ExpressionNode, df: pd.DataFrame) -> pd.Series:
+        """Evaluate logical AND with short-circuit semantics."""
+        if len(node.children) < 2:
+            raise ExpressionError(
+                "AND operation requires at least 2 operands",
+                node_id=node.node_id
+            )
```

Ruff (0.14.10)

Avoid specifying long messages outside the exception class

(TRY003)